src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb

```diff
@@ -887,19 +887,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                                                symbol As Symbol) As Boolean
             Debug.Assert(Me.DeclaringCompilation.EventQueue IsNot Nothing)
 
+            Dim change As Boolean = False
             SyncLock _diagnosticLock
-                Dim change = (variable And mask) = comparand
+                change = (variable And mask) = comparand
                 If change Then
-                    Me.DeclaringCompilation.SymbolDeclaredEvent(symbol)
-
                     If Not ThreadSafeFlagOperations.Set(variable, mask) Then
                         ' If this gets hit, then someone wrote to variable without going through this
                         ' routine, which is a bug.
                         Throw ExceptionUtilities.Unreachable
                     End If
                 End If
-                Return change
             End SyncLock
+
+            If change Then
+                ' Invoke the symbol declared event outside the lock.
+                ' Task scheduler might decide to inline the event processing, which can call back into analyzers,
```

Comment:
The call stack shows that [SymbolDeclaredEvent](http://source.roslyn.io/Microsoft.CodeAnalysis.VisualBasic/R/d7a2310fa51fca81.html) invokes [EnqueuCore](http://source.roslyn.io/Microsoft.CodeAnalysis/R/329bc4b8e37cbd7f.html), which contains a  [SetResult](http://referencesource.microsoft.com/mscorlib/R/810e0ddcb0511ff1.html) invocation on a task completion source, which eventually ends up invoking [FinishContinuations](http://referencesource.microsoft.com/mscorlib/R/dadbf616e42976ee.html), which starts the worker task [ProcessCompilationEventsCoreAsync](http://source.roslyn.io/Microsoft.CodeAnalysis/R/39f4dbd78c629150.html) to process the event and execute analyzers.